### ISSUE TITLE
Feature request: add SmoothMax (LogSumExp)

### ISSUE THREAD

**Issue by AleksPeshkov** [jw1912/bullet#534]
_Comment: AI generated text below provided to make sure I correctly describe what I need. If changing the mainstream code base is not acceptable, I need help how to add this code into generic `simple.rs` file._ for training in CPU.

Standard MAX operations introduce non-differentiable points and zero out gradients for all non-leader elements (sparse gradient problem). This makes it highly unstable for gradient descent when optimizing advanced custom NNUE reduction architectures.

Solution: Implementing a SmoothMax layer via the numerically stable Log-Sum-Exp approach. Its derivative is mathematically equivalent to Softmax, which ensures smooth, dense gradient propagation across all input signals during CPU training batches.

Implementation Details: The proposed code pre-allocates and caches the Softmax weights during the forward pass to achieve maximum efficiency and zero redundant exponential calculations during the manual backward pass execution loop.

```rust
/// A memory-safe, CPU-optimized implementation of the Smooth Maximum (Log-Sum-Exp) 
/// activation/operation node designed for the Bullet NNUE training framework.
/// 
/// This layer provides a fully differentiable, smooth approximation of the `MAX` function, 
/// eliminating the sparse gradient problem ("dead gradients") inherent to standard HardMax.
pub struct BulletSmoothMax {
    /// Caches the calculated Softmax probability distribution during the forward pass.
    /// Used directly as the local gradient vector during the backward pass.
    softmax_cache: Vec<f32>,
}

impl BulletSmoothMax {
    /// Creates a new instance of the SmoothMax operation layer.
    pub fn new() -> Self {
        Self {
            softmax_cache: Vec::new(),
        }
    }

    /// FORWARD PASS
    /// Computes the smooth maximum value of the input vector `s`.
    /// Uses the Log-Sum-Exp mathematical stabilization trick to prevent numerical overflow.
    pub fn forward(&mut self, s: &[f32]) -> f32 {
        if s.is_empty() {
            return 0.0;
        }

        // 1. Find the hard maximum value for numerical stabilization.
        // Shifting inputs by max_val guarantees that exponents are <= 0.0,
        // eliminating the risk of f32::exp resulting in Infinity or NaN.
        let max_val = s.iter().copied().fold(f32::NEG_INFINITY, f32::max);

        // 2. Compute the sum of the shifted exponentials: Sum( exp(s_i - max_val) )
        let mut sum_exp = 0.0;
        for &x in s {
            sum_exp += (x - max_val).exp();
        }

        // 3. Compute the final LogSumExp value: max_val + ln( sum_exp )
        let smooth_max_value = max_val + sum_exp.ln();

        // 4. CACHE LOCAL GRADIENTS FOR BACKWARD PASS
        // The mathematical derivative of LogSumExp with respect to input s_i is exactly Softmax.
        // We compute and store these weights now to avoid redundant operations later.
        self.softmax_cache.resize(s.len(), 0.0);
        for (i, &x) in s.iter().enumerate() {
            // Formula: exp(s_i - max) / sum( exp(s_j - max) )
            self.softmax_cache[i] = (x - max_val).exp() / sum_exp;
        }

        smooth_max_value
    }

    /// BACKWARD PASS
    /// Performs backpropagation across the operation node using the Chain Rule.
    /// 
    /// # Arguments
    /// * `incoming_gradient` - The global gradient backpropagated from subsequent layers (upstream).
    /// * `output_gradients` - The mutable buffer where calculated gradients for each input element `s_i` are accumulated.
    pub fn backward(&self, incoming_gradient: f32, output_gradients: &mut [f32]) {
        assert_eq!(
            self.softmax_cache.len(), 
            output_gradients.len(), 
            "The size of the output gradients buffer must match the internal cache size!"
        );

        // Chain Rule application: 
        // dLoss/ds_i = (dLoss/dSmoothMax) * (dSmoothMax/ds_i) = incoming_gradient * softmax_weight
        for (out_grad, &local_softmax_weight) in output_gradients.iter_mut().zip(&self.softmax_cache) {
            *out_grad += incoming_gradient * local_softmax_weight;
        }
    }
}
```

**Comment by jw1912** [comment]
There is no CPU training on main. Adding this to main for GPU is quite simple so I could do that. I assume you’re using some ancient version of bullet that still has CPU training - in which case I am not really interested in supporting such old versions, you can look through the source code (or use an LLM) to see how to add custom operations.

Also what are you planning to use this for?

**Comment by AleksPeshkov** [comment]
I am working on non SCRELU/CRELU net architecture with idea to hold native chess symmetry around 0. Particularly I explore properties of SUM(N)*ABS(MAX(N)) many to one reduction (alone or together other multiplicators).

I do not want to bother with GPU drivers for AMD CPU. I doubt they even exist.
